[GILJOB-1] 프로젝트 CI 구성 및 테스트 환경 설정

raejoonee opened the pull request:
## Summary
* Jest를 사용한 프로젝트 테스트 환경을 설정했습니다.
* GitHub Actions를 사용하여 CI를 만들었습니다.

## Detail
* 앞으로 `.github/workflows/CI.yml`이 존재하는 모든 브랜치에서 push가 발생할 때마다, **프로젝트 빌드 성공 여부**와 **프로젝트 내부의 모든 테스트 케이스의 성공 여부**에 따라 merge가 가능하거나 불가능해집니다.
* Create React App을 사용하지 않았기 때문에 tsx 파일 첫 줄에 **`import React from 'react'`가 필수**입니다. 😢 향후 개발 시 유의해주세요!! 💡 _Webpack 설정을 통해 이 문제를 해결하려 했으나 Jest에서 계속 오류가 발생하여 `import React from 'react'`를 쓰는 것으로 결정됨_

raejoonee committed:
[GILJOB-21] env: GitHub Actions를 사용하여 CI pipeline의 일부를 구성한다.

raejoonee committed:
[GILJOB-27] env: Jest를 사용하기 위해 의존 모듈을 설치하고 Jest 개발 환경을 구성한다.

raejoonee committed:
[GILJOB-27] test: 테스트 컴포넌트가 정상적으로 렌더링되는지 테스트한다.

raejoonee committed:
[GILJOB-27] fix: Jest 환경을 위해서 React를 다시 import하도록 수정한다.

raejoonee committed:
[GILJOB-21] env: CI 과정에 테스트를 추가한다.

kyoung-jnn commented on jest.config.js:
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  setupFilesAfterEnv: [
+    'given2/setup',
+    'jest-plugin-context/setup',
+  ],
+  testEnvironment: 'jsdom',
+};
```
`babel-test` 설치해서 es6+ 대비해야하지 않을까요?
```
transform: {
        '^.+\\.(ts|tsx)?$': 'babel-jest'
},
```
jest 는 잘 몰라서 공부하면서 보고있습니다..
https://jestjs.io/docs/getting-started#using-babel

raejoonee replied:
주말에 팀 미팅 결과 없어도 문제 없을 것으로 판단되어 해결되었습니다!

kyoung-jnn commented on src/App.test.tsx:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import { render } from '@testing-library/react';
+import '@testing-library/jest-dom';
```
`jest-setup.js` 파일로 `import '@testing-library/jest-dom'`문 따로 빼서
`jest.config.js` 파일 `setupFilesAfterEnv` 옵션에
`setupFilesAfterEnv: ["<rootDir>/jest-setup.js"]` 추가하면 굳이 import 안해도 됩니다!

https://stackoverflow.com/questions/65947682/expect-tohaveattribute-is-not-a-function-why

raejoonee replied:
헉 😮  감사합니다!! 한 번 시도해 보겠습니다.

raejoonee replied:
이대로 시도했음에도 컴포넌트에 `import React from 'react'` 없이는 test가 제대로 진행되지 않네요 😢

kyoung-jnn commented on src/test-utils.tsx:
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import { render as tlRender } from '@testing-library/react';
+
+const render = (component: React.ReactElement) => tlRender(component);
+
+export * from '@testing-library/react';
+export { render };
```
이 파일은 생성은 되어 있는데 사용은 안되는거 같아서 어디에 사용되나요? 🤔
`jest.config.js`의 `setupFilesAfterEnv` 옵션에 사용되나여?

raejoonee replied:
Redux를 프로젝트에 도입할 때 최상위 컴포넌트를 `<Provider>`로 감싸야 하는데, 컴포넌트 단위 테스트를 하려면 이 `<Provider>`를 매번 감싸야 하는 불편함이 있기 때문에 이미 `<Provider>`로 감싼 테스트 유틸을 만들고 이 유틸을 테스트할 때 import하여 사용하려고 만든 겁니다! 이런 방식은 [Redux 공식 문서](https://redux.js.org/usage/writing-tests#components)에서 권장하는 컴포넌트 단위 테스트 기법입니다.

LeeJuhae reviewed (approved):
Jest 테스트 환경 설정과 Github actions를 이용한 CI pipeline 구성하신 것 잘 봤습니다!

raejoonee committed:
fix: jest 설정을 수정한다.

https://github.com/YAPP-19th/Web-Team-1-Frontend/pull/2#discussion_r721139568
https://stackoverflow.com/questions/65947682/expect-tohaveattribute-is-not-a-function-why

raejoonee committed:
fix: App.tsx에서 React를 import하지 않도록 수정한다.

raejoonee committed:
fix: App.tsx에서 React를 import하도록 수정한다.

raejoonee committed:
Merge branch 'env/base' into env/ci

raejoonee committed:
Fix: package-lock.json 파일을 수정한다.
